### PR TITLE
Add layers to cluster graphs, improve some doc strings, expand tests

### PR DESCRIPTION
**Pull request by bannanc**
In this pull request I've add code to add atoms without smirks-indices to cluster graphs. 
It seems to be working the way I would want so far, but I've only tested a few cases, I would like to test more and update the relevant example notebook before we merge.

@davidlmobley or @jmaat if you have time to take a look that would be great.

### CODE REVIEW

File: chemper/graphs/cluster_graph.py [hunk 1]
```diff
@@ -100,6 +105,36 @@ def add_atom(self, atom):
...
+            """
+
+            Parameters
+            ----------
+            atom: chemper Atom
+
+            Returns
+            -------
+            score: int
+                 how similar is atom to current storage, max of 7 for all decorators identical
```
Comment (davidlmobley) [inline chemper/graphs/cluster_graph.py]: "How similar is atom to current storage" is opaque to me. Can you clarify? Does this mean basically out of the total number of possible decorators (7?) how many of these are the same? (And it is only applied when the element is already the same?)

File: chemper/graphs/cluster_graph.py [hunk 2]
```diff
@@ -100,6 +105,36 @@ def add_atom(self, atom):
...
+                 0 if atom's atomic number not included in current set
```
Comment (davidlmobley) [inline chemper/graphs/cluster_graph.py]: "0 if atom's atomic number..." also not totally clear. Does this mean that if you have a wildcard atom but with similar decorators it is still 0?

File: chemper/graphs/cluster_graph.py [hunk 3]
```diff
@@ -100,6 +105,36 @@ def add_atom(self, atom):
...
+            """
+            score = 0
+            decs = self.make_atom_decorators(atom)
+            atomic_nums = [d[0] for d in self.decorators]
```
Comment (davidlmobley) [inline chemper/graphs/cluster_graph.py]: What is `d[0]` if it is a wildcard? I'm unclear what this does for wildcards.

File: chemper/graphs/cluster_graph.py [hunk 4]
```diff
@@ -262,6 +297,128 @@ def _add_first_smirks_atoms(self, mol, smirks_atoms):
                                          neighbor_storage,
                                          bond=bond_storage)
 
+        for smirks_index, atom_index in smirks_atoms.items():
```
Comment (davidlmobley) [inline chemper/graphs/cluster_graph.py]: Seems like you should have comments here; not totally obvious what you're doing without thinking about it for a while/reading context.

File: chemper/graphs/cluster_graph.py [hunk 5]
```diff
@@ -262,6 +297,128 @@ def _add_first_smirks_atoms(self, mol, smirks_atoms):
...
+            atom = mol.get_atom_by_index(atom_index)
+            storage = self.atom_by_smirks_index[smirks_index]
+            self._add_layers(mol, atom, storage, self.layers, atom_dict)
+
+    def _add_layers(self, mol, atom, storage, layers, idx_dict):
```
Comment (davidlmobley) [inline chemper/graphs/cluster_graph.py]: This part is also probably involved enough you should have comments that briefly explain the algorithm to make the code easier to follow.

File: chemper/graphs/fragment_graph.py [hunk 6]
```diff
@@ -152,13 +152,15 @@ def as_smirks(self):
             return None
 
         if self.atom_by_smirks_index:
-            min_smirks = min(self.atom_by_smirks_index)
+            # sometimes we use negative numbers for internal indexing
```
Comment (davidlmobley) [inline chemper/graphs/fragment_graph.py]: Why would negative indices be used?

File: chemper/graphs/fragment_graph.py [hunk 7]
```diff
@@ -271,7 +273,7 @@ def add_atom(self, new_atom, new_bond=None, bond_to_atom=None,
 
         new_atom_storage = self.AtomStorage(new_atom, smirks_index=new_smirks_index)
         self._graph.add_node(new_atom_storage)
-        if new_smirks_index is not None and new_smirks_index > 0:
+        if new_smirks_index is not None:
```
Comment (davidlmobley) [inline chemper/graphs/fragment_graph.py]: You removed a check that only does this if the index is greater than 0; deliberate? Why? (Just confirming.)

File: tests/test_basic_graphs.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+    newer_atom = c.add_atom(None, None, new_atom)
+    assert c.as_smirks() == '[*]~[*]'
+
+    c = graph_method()
+    c.add_atom(None, new_smirks_index=1)
+    assert c.as_smirks() == '[*:1]'
+
+
+# make sure corner cases can at least be built without failing
```
Comment (davidlmobley) [inline tests/test_basic_graphs.py]: Why are these corner cases?

File: tests/test_graphs_output.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+from chemper.graphs.cluster_graph import ClusterGraph
+from chemper.graphs.fragment_graph import ChemPerGraphFromMol, ChemPerGraph
+from chemper.mol_toolkits import mol_toolkit
+import pytest
+
+
+def make_frag_graph(smiles, layers):
+    """
+    Generates a chemper Mol from the provided smiles
+    a ChemPerGraphFromMol with that molecules with the layers specified
```
Comment (davidlmobley) [inline tests/test_graphs_output.py]: The description of this test does not seem comprehensible.

File: chemper/graphs/cluster_graph.py (hunk 1 above)
Comment (bannanc) [reply chemper/graphs/cluster_graph.py]: Yes, your intuition was correct, but I will update the wording so that is more clear.

File: chemper/graphs/cluster_graph.py (hunk 2 above)
Comment (bannanc) [reply chemper/graphs/cluster_graph.py]: So I hadn't thought through the "wildcard" option completely. The input is always an atom, but its true that technically the atom storage (self) could be empty, which right now would cause a problem, I think it makes most sense to give that a high score since if the storage is empty it should match with any atom. I need to think about the best way to handle that scenario.

File: chemper/graphs/cluster_graph.py (hunk 4 above)
Comment (bannanc) [reply chemper/graphs/cluster_graph.py]: I was going to add more detail to the new functions, but this one could probably use more help too.

File: chemper/graphs/fragment_graph.py (hunk 6 above)
Comment (bannanc) [reply chemper/graphs/fragment_graph.py]: I'm using negative indices to track the location of non-indexed atoms. ChemicalEnvironments do this internally too where alpha or beta atoms are located by their negative smirks index. I probably need to add more information in the doc strings on what those values are allowed to be.

File: chemper/graphs/fragment_graph.py (hunk 7 above)
Comment (bannanc) [reply chemper/graphs/fragment_graph.py]: yes, it was deliberate. Originally I was thinking of `atom_by_smirks_index` for only atoms that are actually tagged with smirks indices used in the smirks pattern (positive values), however, I realized for the cluster graph that using unique smirks indices made a convenient way to book keep atoms so I wanted the dictionary to have all of them.

File: tests/test_basic_graphs.py (hunk 8 above)
Comment (bannanc) [reply tests/test_basic_graphs.py]: Corner cases was maybe not the best word for them. They were mostly added to increase coverage. Now that I think about it the `test_no_fail_fragment` could probably also have expected smirks patterns for at least the 1 layer options, so maybe those should go above. 

The multiple layers for the cluster graphs become really really complicated to write out and because decorators are stored as sets and not lists the SMIRKS patterns aren't consistent, I could maybe make a sorting option that makes the smirks patterns consistent so that more intense testing can be done instead of just "does it not fail"

File: chemper/graphs/fragment_graph.py (hunk 6 above)
Comment (bannanc) [reply chemper/graphs/fragment_graph.py]: I'm going to rename this as label, with details that positive labels are included in the SMIRKS patter, but that all other labels are just used as identification for storage.

**Comment by bannanc** [comment]
I'm going to go ahead and merge this PR. 
I've addressed all the comments David brought up before. 
I didn't end up adding a new notebook example, but I'd like to finish updating the docs and current examples so anything I add is more clearly different.
